### PR TITLE
JAMES-2227 System mailboxes should be case sensitive

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Role.java [hunk 1]
```diff
@@ -41,12 +41,12 @@
...
-    public static final Role ARCHIVE = new Role("archive");
-    public static final Role SPAM = new Role("spam");
-    public static final Role TEMPLATES = new Role("templates");
+    public static final Role ARCHIVE = new Role("archive", DefaultMailboxes.ARCHIVE);
+    public static final Role SPAM = new Role("spam", DefaultMailboxes.SPAM);
+    public static final Role TEMPLATES = new Role("templates", DefaultMailboxes.TEMPLATES);
     
     private static final Map<String, Role> ROLES = 
-            ImmutableList.<Role>of(INBOX, ARCHIVE, DRAFTS, OUTBOX, SENT, TRASH, SPAM, TEMPLATES)
+            ImmutableList.<Role>of(INBOX, DRAFTS, OUTBOX, SENT, TRASH, ARCHIVE, SPAM, TEMPLATES)
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Role.java]: why did you change the ordering ? Why don't you sort alphabetically to make it easy for a human to search in it ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java [hunk 2]
```diff
@@ -208,7 +208,7 @@ public void getMailboxesShouldReturnSortOrder1000OnUnknownFolder() throws Except
     }
 
     @Test
-    public void getMailboxesShouldReturnInboxWithSortOrder10OnDifferenceCase() throws Exception {
+    public void getMailboxesShouldReturnCommonMailboxSortOrderWhenInboxOnDifferenceCase() throws Exception {
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java]: I don't agree to change this test. It is to make sure Inbox always appear before other mailboxes

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java [hunk 3]
```diff
@@ -20,39 +20,17 @@
...
         assertThat(Role.from("jjjj")).isEqualTo(Optional.empty());
     }
 
-    @Test
-    public void fromShouldReturnInboxWhenContainsUppercaseValue() {
-        assertThat(Role.from("InBox")).isEqualTo(Optional.of(Role.INBOX));
-    }
-
-    @Test
-    public void fromShouldReturnInboxWhenContainsUppercaseValueInTurkish() {
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java]: you lost this very good test case

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Role.java (hunk 1 above)
Comment (aduprat) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Role.java]: To match the order of the constant just the lines above.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java (hunk 2 above)
Comment (aduprat) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java]: It's not _Inbox_ in this test, but _InBox_

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java (hunk 3 above)
Comment (aduprat) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java]: We do not need this anymore, we are never converting the string case.

**Comment by rouazana** [comment]
Sorry to not have been part of this discussion before, but this is wrong (rfc3501):
> mailbox         = "INBOX" / astring
                    ; INBOX is case-insensitive.  All case variants of
                    ; INBOX (e.g., "iNbOx") MUST be interpreted as INBOX
                    ; not as an astring.  An astring which consists of
                    ; the case-insensitive sequence "I" "N" "B" "O" "X"
                    ; is considered to be INBOX and not an astring.
                    ;  Refer to section 5.1 for further
                    ; semantic details of mailbox names.

**Comment by rouazana** [comment]
Also:
>    The case-insensitive mailbox name INBOX is a special name reserved to
   mean "the primary mailbox for this user on this server".  The
   interpretation of all other names is implementation-dependent.

  > In particular, this specification takes no position on case
   sensitivity in non-INBOX mailbox names.  Some server implementations
   are fully case-sensitive; others preserve case of a newly-created
   name but otherwise are case-insensitive; and yet others coerce names
   to a particular case.  Client implementations MUST interact with any
   of these.  If a server implementation interprets non-INBOX mailbox
   names as case-insensitive, it MUST treat names using the
   international naming convention specially as described in section
   5.1.3.

For my part I prefer to be case insensitive for every system mailbox, to be consistent with INBOX, but we can discuss this.

**Comment by aduprat** [comment]
After doing the code modifications, I agree with you, it seems more dangerous & we lack the INBOX case.

I will propose with another solution.

**Comment by aduprat** [comment]
The other solution is here:
https://github.com/linagora/james-project/pull/1143

**Comment by aduprat** [comment]
With correct INBOX case implementation.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java (hunk 3 above)
Comment (chibenwa) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java]: Why can't we restore this test in the final diff?

Comment (aduprat) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java]: Ok, I will add it; but it has no real added value as we are no more changing the case of the string.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java (hunk 2 above)
Comment (mbaechler) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java]: and inbox is always case insensitive

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Role.java [hunk 4]
```diff
@@ -35,43 +33,53 @@
 public class Role {
 
     public static final String USER_DEFINED_ROLE_PREFIX = "x-";
+
+    private static final BiFunction<String, String, Boolean> SENSITIVE_COMPARATOR = (a, b) -> a.equals(b);
+    private static final BiFunction<String, String, Boolean> NON_SENSITIVE_COMPARATOR = (a, b) -> a.equalsIgnoreCase(b);
+
+    public static final Role INBOX = new Role("inbox", DefaultMailboxes.INBOX, NON_SENSITIVE_COMPARATOR);
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Role.java]: s/SENSITIVE/CASE_SENSITIVE/ everywhere

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java [hunk 5]
```diff
@@ -225,20 +225,20 @@ public void getMailboxesShouldReturnInboxWithSortOrder10OnDifferenceCase() throw
                 .extracting(GetMailboxesResponse.class::cast)
                 .flatExtracting(GetMailboxesResponse::getList)
                 .extracting(Mailbox::getSortOrder)
-                .containsOnly(SortOrder.of(10));
+                .containsOnly(SortOrder.of(1000));
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java]: as InBoX should be the same than Inbox, this tests should not be changed anymore, no?
